### ISSUE TITLE
Improve torch.nn documentation loading time

### ISSUE THREAD

**Issue by DomHudson** [pytorch/pytorch#20984]
## 📚 Documentation
Hi, this is not a serious issue but can be a little frustrating. I've noticed the documentation for the `nn` package (https://pytorch.org/docs/stable/nn.html) is quite slow to load. It takes around 16 seconds to render the page on Google Chrome. Often, I am looking for the documentation for a specific class and chrome's search won't find it until the page is done rendering.

![Screenshot from 2019-05-27 10-55-22](https://user-images.githubusercontent.com/10864294/58412256-fd3dff00-806d-11e9-8754-d80c1041e2f8.png)

## Related entities
https://github.com/pytorch/pytorch/issues/11968

**Comment by Jonas1312** [comment]
Takes around 11s to load `torch.nn` doc on my dual core cpu 😕

**Comment by jeremycochoy** [comment]
Let's be honest. More than 5s for a documentation, this is a serious issue.

**Comment by ezyang** [comment]
An bandaid is to start splitting up nn into multiple pages. But I think people like being able to full text search on all nn modules. So someone would need to figure out why it takes so long to render (is it math equations). Maybe a stopgap is to make another version of the docs with very little CSS / layout formatting. Another stopgap might be to revert to the old docs style, which apparently were less affected by this problem.

**Comment by Kaixhin** [comment]
Just had a quick look at this. For maths it's using KaTeX, which is supposed to be faster than MathJax, but I'm not sure it's using server-side rendering. There's a bunch of JS libraries like jQuery and underscore that are being loaded from the website instead of a CDN. Finally there's some data being requested by `theme.js` that's being requested significantly after the rest of the resources.

**Comment by ezyang** [comment]
@zou3519 says it will be hard to get this to render on server side; we'll have to add image generation during compile time.

**Comment by ezyang** [comment]
@brianjo responded to me:

> We reached out the Shift about the performance issues we were seeing with torch.nn in the documentation and they basically told us it was due to the size of the content and there wasn’t much they could do about improving the performance on the theme side. They also mentioned that performance wasn’t that different between the old theme and the new one, so changing that back wouldn’t help that much.
>
> They did see a performance gain between our pages on Master vs Stable (Where we broke the page into three sections.)  We could improve that further if we broke the pages into more sections.
>
> So the questions for you are:
>
> * Is the new breakdown on Master performant enough?  - If so we should make those changes on Stable or just wait for 1.2 release in a 5  weeks.
> * If it’s not performant enough, should we break it down further? – If so, I could use some help determining how to best do this.

I don't think it's fast enough. https://pytorch.org/docs/master/nn.html still takes 14s before the the "load spinner" stops.

I think it is pretty clear that the problem is KaTeX. To show this, take the source code for nn.html and delete the script tags which load KaTeX. Load time immediately drops to 3s, which is *much* more reasonable. In that case, we should replace KaTeX with prerendered server images, rather than splitting the page further.

**Comment by brianjo** [comment]
Thanks Edward! I’ll take a look at optimizing from that angle.

Cheers,

Brian

**Comment by zou3519** [comment]
I don't think that using katex is the way to go. I took a quick look at the "sphinxcontrib-katex" plugin that we're using. From what I understand it doesn't support server-side rendering, only client-side. katex by itself does have this ability, but we'd need to be running node.js and we'd need a way to reach into sphinx and ask it for things that need to be rendered. If we want to go with the katex route, we'd probably need a new sphinx plugin / integration.

However, there are some sphinx-blessed ways to pre-generate images. @brianjo [This looks pretty promising](https://shimizukawa-sphinx.readthedocs.io/en/latest/ext/math.html#module-sphinx.ext.imgmath), but it means that the docs need a dependency on latex and dvipng and I'm not sure how hard those are to install in general.

**Comment by oguzserbetci** [comment]
I think the page can be broken up to at least into: `modules`, `functional`, `losses`, `init`, `utilities`. The slowness makes the whole docs very unusable. It also effects doc lookup programs like dash that use directly the html site.

**Comment by ezyang** [comment]
Apparently, not only do we use katex, we also have mathjax enabled on the underlying theme.

**Comment by ezyang** [comment]
OK, some updates:

- In #23025 I trade off network performance for CPU performance: with the removal of KaTeX, the amount of CPU needed by the docs page is greatly reduced, but now I'm forcing your browser to make lots of itty bitty image calls. On a fast network, these seem to load pretty quick, and get subsequently cached, so I *think* this is an improvement, but I'm not 100% sure.
- When I was testing by emulating a slow 3G connection, a bottleneck with images is sending a pile of HTTP requests to get 304 back from the server saying nothing has changed; this is because Chrome by default will only open N connections at any given point in time. On the live documentation server we set a fairly short cache window (something like less than a day), which means you'll be reloading these resources a lot. We should probably set a longer cache time for these resources.
- After switching to images, most CPU is spent repeatedly repainting nodes when images finish loading. Luckily, there isn't a cascading effect; most repaints only affect seven or so nodes, but it takes up the bulk of processing time when you slow down the CPU. In principle we could avoid the repaint by correctly sizing the images when we initially include them, but Sphinx doesn't seem to support it. (Does feel like it would be relatively easy to add.)

**Comment by ezyang** [comment]
Those of you who are affected by this: which of these pages loads most quickly for you?

* https://upbeat-roentgen-bdeb9e.netlify.com/ - imgmath with png
* https://5d39ea246de708000ddda861--upbeat-roentgen-bdeb9e.netlify.com/ - prerendered KaTeX
* https://5d39ed69d7ddde00084f8c12--upbeat-roentgen-bdeb9e.netlify.com/ - imgmath svg

**Comment by Jonas1312** [comment]
3rd one (imgmath svg)

**Comment by Chillee** [comment]
Average of 3 after first load (using cache) (and a random load graph)
> * https://upbeat-roentgen-bdeb9e.netlify.com/nn.html#loss-functions - imgmath with png

4.63s

![image](https://user-images.githubusercontent.com/6355099/61903123-a0798c00-aed8-11e9-94cc-68add640a097.png)


> * https://5d39ea246de708000ddda861--upbeat-roentgen-bdeb9e.netlify.com/nn.html#loss-functions - prerendered KaTeX

7.10s

![image](https://user-images.githubusercontent.com/6355099/61903205-d4ed4800-aed8-11e9-96c0-b96048fe1b64.png)


> * https://5d39ed69d7ddde00084f8c12--upbeat-roentgen-bdeb9e.netlify.com/nn.html#loss-functions - imgmath svg

5.20s

![image](https://user-images.githubusercontent.com/6355099/61903270-f3ebda00-aed8-11e9-8e3d-2a108bbe536e.png)


> * https://pytorch.org/docs/master/nn.html#loss-functions - non-prerendered KaTeX (at time of writing)

9.43s (this one also had the most variance)

![image](https://user-images.githubusercontent.com/6355099/61903378-37dedf00-aed9-11e9-90a4-d9d6aca04f02.png)


Google Chrome, i7-8750H, Thinkpad X1E

I'll post some data for cacheless load and with my mac later.

**Comment by Chillee** [comment]
For Macbook (no averages since too much work):


> * https://upbeat-roentgen-bdeb9e.netlify.com/nn.html#loss-functions - imgmath with png

4.65

> * https://5d39ea246de708000ddda861--upbeat-roentgen-bdeb9e.netlify.com/nn.html#loss-functions - prerendered KaTeX

6.9

> * https://5d39ed69d7ddde00084f8c12--upbeat-roentgen-bdeb9e.netlify.com/nn.html#loss-functions - imgmath svg

15.26!
> * https://pytorch.org/docs/master/nn.html#loss-functions - non-prerendered KaTeX (at time of writing)

**Comment by ezyang** [comment]
@Chillee Is that 15.26! for svg, or non-prerendered KaTeX?

**Comment by ezyang** [comment]
@Jonas123 @Chillee Could you also check https://5d3a0845dc1dcc0007252ba2--upbeat-roentgen-bdeb9e.netlify.com/nn.html#conv2d as well? This one is prerendered KaTeX but I killed the CSS and JS on page and added a bunch of preloading.

**Comment by brianjo** [comment]
I ran these through [Google PageSpeed](https://developers.google.com/speed/pagespeed/insights/)
![Screenshot 2019-07-25 12 47 35](https://user-images.githubusercontent.com/1500149/61904620-0d425580-aedc-11e9-9726-d95424e1a653.png)
![Screenshot 2019-07-25 12 41 50](https://user-images.githubusercontent.com/1500149/61904621-0d425580-aedc-11e9-9cd0-4ded0689114a.png)
![Screenshot 2019-07-25 12 42 37](https://user-images.githubusercontent.com/1500149/61904622-0d425580-aedc-11e9-92c4-1ea2f1f7dd97.png)
![Screenshot 2019-07-25 12 41 24](https://user-images.githubusercontent.com/1500149/61904624-0ddaec00-aedc-11e9-836c-466d4009d2ad.png)
![Screenshot 2019-07-25 12 58 29](https://user-images.githubusercontent.com/1500149/61904625-0ddaec00-aedc-11e9-8131-70d2c5b6a0e8.png)

**Comment by ezyang** [comment]
Thanks @brianjo. I'm afraid that PageSpeed doesn't give the full picture: it likes the prerendered KaTeX the best, but as the other folks here has mentioned, it doesn't actually perform all that well.

**Comment by Chillee** [comment]
@ezyang that was the prerendered svg, but it's possible I was doing something weird (I was rushing to lunch). I'll check it again when I get back.

**Comment by ezyang** [comment]
@Chillee I have noticed that sometimes, the SVG renderer stalls, and it takes a lot longer for the svg to render (Macbook Pro, Chrome). It doesn't happen consistently though.

**Comment by Chillee** [comment]
@ezyang yeah when trying it again it seemed pretty normal (on par with the png times).

The new page (with no HTML/JS) loads in ~3.7s on mac and about ~4.0s on my linux laptop.

**Comment by ezyang** [comment]
@Chillee  Thanks. If all else fails, we'll go SVG, but I *do* think the KaTeX looks better, so I am pretty interested if we can get it fast enough. Here's my latest attempt: https://5d3a1189e90ce1000780e247--upbeat-roentgen-bdeb9e.netlify.com/nn.html#pixelshuffle (most of the change is preloading webfonts for KaTeX and the website.) This is pretty snappy on my browser; I don't know about yours though.

**Comment by Chillee** [comment]
I do agree that the KaTeX looks better, and the newest attempt takes about ~4s on my Mac, and about ~5.5s on my Linux laptop.

**Comment by ezyang** [comment]
That sounds better than imgmath svg/png :)

**Comment by Chillee** [comment]
It does indeed :D

**Comment by Chillee** [comment]
Don't know if you've been messing around with it, but the link anchors no longer work on the most recent one: https://5d3a1189e90ce1000780e247--upbeat-roentgen-bdeb9e.netlify.com/nn.html#pixelshuffle

**Comment by ezyang** [comment]
Yes, that will need to get fixed. I guess there was some JavaScript that was making them visible; we should switch that to CSS only.

**Comment by cgarciae** [comment]
The "single page per module" strategy makes it VERY HARD to search for Pytorch documentation since Google will just redirect you to the top of the page, then you have to hit `Ctrl + f` and type the term again. 

Why not break it into a page per function / class like most other libraries (numpy, pandas, sklearn, etc)? I would reduce the load time problem discussed here by virtue of just having to process less content and it would make its content easier to find.

**Comment by zou3519** [comment]
@cgarciae https://github.com/pytorch/pytorch/issues/32838 is the tracking issue for the Google search problem, it would be great to also voice your support there for more visibility

**Comment by vadimkantorov** [comment]
@ezyang One way is also supporting both things: page-per-function for super-fast redirection/rendering for leads from google search; and a giant optimized (when it happens) single page (or page per module)

**Comment by zou3519** [comment]
@vadimkantorov we're splitting up the nn documentation in #34796

**Comment by vadimkantorov** [comment]
@zou3519 If I understand well in there splitting is done by modules, but maybe still a page per function is good for instant google->docs transition

**Comment by zou3519** [comment]
I agree, the other frameworks (numpy, Tensorflow) have all adopted a single page per function

**Comment by vadimkantorov** [comment]
I think all functions on one page is good for roaming / discoverability by random scrollong, and function per page is good for hyperlinks and google, so both are sensible to have

**Comment by Enealor** [comment]
I like breaking up the documentation as it will make it faster to build small parts of it - any changes to documentation included in nn.rst is slow to process due to how long it takes to compile. If we want to use one page per function/class, then we should look into using [autosummary](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html). It handles generating the individual HTML pages and creating links to them.
